AppCenterAnalytics/AppCenterAnalytics/Internals/Model/MSEventLog.m

```diff
@@ -68,53 +82,114 @@ - (MSCommonSchemaLog *)toCommonSchemaLogForTargetToken:(NSString *)token {
+  csLog.ext.metadataExt = [MSMetadataExtension new];
+  [self setPropertiesAndMetadataForCSLog:csLog];
   return csLog;
 }
 
 #pragma mark - Helper
 
-- (NSDictionary<NSString *, NSObject *> *)convertACPropertiesToCSproperties:(NSDictionary<NSString *, NSString *> *)acProperties {
+- (void)setPropertiesAndMetadataForCSLog:(MSCommonSchemaLog *)csLog {
   NSMutableDictionary *csProperties;
-  if (acProperties) {
+  NSMutableDictionary *metadata;
+  if (self.typedProperties) {
     csProperties = [NSMutableDictionary new];
-    for (NSString *acKey in acProperties) {
+    metadata = [NSMutableDictionary new];
+    for (MSTypedProperty *typedProperty in [self.typedProperties.properties objectEnumerator]) {
 
       // Properties keys are mixed up with other keys from Data, make sure they don't conflict.
-      if ([acKey isEqualToString:kMSDataBaseData] || [acKey isEqualToString:kMSDataBaseDataType]) {
-        MSLogWarning(MSAnalytics.logTag, @"Cannot use %@ in properties, skipping that property.", acKey);
+      if ([typedProperty.name isEqualToString:kMSDataBaseData] || [typedProperty.name isEqualToString:kMSDataBaseDataType]) {
+        MSLogWarning(MSAnalytics.logTag, @"Cannot use %@ in properties, skipping that property.", typedProperty.name);
         continue;
       }
+      [self addTypedProperty:typedProperty toCSMetadata:metadata AndCSProperties:csProperties];
+    }
+  }
+  if (csProperties.count != 0) {
+    csLog.data.properties = csProperties;
+  }
+  if (metadata.count != 0) {
+    csLog.ext.metadataExt.metadata = metadata;
+  }
+}
 
-      // If the key contains a '.' then it's nested objects (i.e: "a.b":"value" => {"a":{"b":"value"}}).
-      NSArray *csKeys = [acKey componentsSeparatedByString:@"."];
-      NSUInteger lastIndex = csKeys.count - 1;
-      NSMutableDictionary *destProperties = csProperties;
-      for (NSUInteger i = 0; i < lastIndex; i++) {
-        NSMutableDictionary *subObject = nil;
-        if ([(NSObject *) destProperties[csKeys[i]] isKindOfClass:[NSMutableDictionary class]]) {
-          subObject = destProperties[csKeys[i]];
-        }
-        if (!subObject) {
-          if (destProperties[csKeys[i]]) {
-            MSLogWarning(MSAnalytics.logTag, @"Property key '%@' already has a value, the old value will be overridden.", csKeys[i]);
-          }
-          subObject = [NSMutableDictionary new];
-          destProperties[csKeys[i]] = subObject;
+- (void)addTypedProperty:(MSTypedProperty *)typedProperty toCSMetadata:(NSMutableDictionary *)csMetadata AndCSProperties:(NSMutableDictionary *)csProperties {
+  NSNumber *typeId = self.metadataTypeIdMapping[typedProperty.type];
+
+  // If the key contains a '.' then it's nested objects (i.e: "a.b":"value" => {"a":{"b":"value"}}).
+  NSArray *csKeys = [typedProperty.name componentsSeparatedByString:@"."];
+  NSMutableDictionary *propertyTree = csProperties;
+  NSMutableDictionary *metadataTree = csMetadata;
+  for (NSUInteger i = 0; i < csKeys.count - 1; i++) {
```

Comment:
Isn't this for loop too complex than necessary?
``` 
   if (![(NSObject *) propertyTree[csKeys[i]] isKindOfClass:[NSMutableDictionary class]]) {
      if (propertyTree[csKeys[i]]) {
        propertyTree = nil;
        break;
      } else {
        propertyTree[csKeys[i]] = [NSMutableDictionary new];
      }
    }

    propertyTree = propertyTree[csKeys[i]];
    if (typeId) {
      if (!metadataTree[kMSFieldDelimiter]) {
        metadataTree[kMSFieldDelimiter] = [NSMutableDictionary new];
      }
      if (!metadataTree[kMSFieldDelimiter][csKeys[i]]) {
        metadataTree[kMSFieldDelimiter][csKeys[i]] = [NSMutableDictionary new];
      }
      metadataTree = metadataTree[kMSFieldDelimiter][csKeys[i]];
    }

Reply:
Can you explain which parts are unnecessary? I believe that snippet misses several edge cases, but I want to be sure

Reply:
The snippet is more like to remove duplicated code blocks. FYI, the snippet passed all unit tests in this branch.

Reply:
Yes looks like we could improve it like this!